plugins/chat/app/services/chat/auto_leave_channels.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module Chat
+  class AutoLeaveChannels
+    include Service::Base
+
+    ALLOWED_GROUP_PERMISSIONS ||= [
+      CategoryGroup.permission_types[:create_post],
+      CategoryGroup.permission_types[:full],
+    ]
+
+    policy :chat_enabled?
+
+    contract do
+      attribute :event
+      attribute :user_id, :integer
+      attribute :group_id, :integer
+      attribute :channel_id, :integer
+      attribute :category_id, :integer
+    end
+
+    step :remove_memberships
+
+    private
+
+    def chat_enabled?
+      SiteSetting.chat_enabled
+    end
+
+    def remove_memberships
+      group_ids = SiteSetting.chat_allowed_groups_map
+      group_permissions = ALLOWED_GROUP_PERMISSIONS
+
+      if !group_ids.include?(Group::AUTO_GROUPS[:everyone])
+        sql = <<~SQL
+          DELETE FROM user_chat_channel_memberships uccm
+          WHERE NOT EXISTS (
+            SELECT 1 
+            FROM group_users gu 
+            WHERE gu.user_id = uccm.user_id 
+            AND gu.group_id IN (:group_ids)
+          )
+          RETURNING chat_channel_id, user_id
+        SQL
+
+        users_removed_map = Hash.new { |h, k| h[k] = [] }
+
+        DB
+          .query_array(sql, group_ids:)
+          .each { |channel_id, user_id| users_removed_map[channel_id] << user_id }
+
+        Chat::Action::PublishAutoRemovedUser.call(event: context.event, users_removed_map:)
+      end
+
+      user_sql = context.user_id.to_i > 0 ? "AND u.id = #{context.user_id}" : ""
+      channel_sql = context.channel_id.to_i > 0 ? "AND cc.id = #{context.channel_id}" : ""
+      category_sql = context.category_id.to_i > 0 ? "AND c.id = #{context.category_id}" : ""
+
+      sql = <<~SQL
+        WITH valid_permissions AS (
+          SELECT gu.user_id, cg.category_id
+          FROM group_users gu
+          JOIN category_groups cg ON cg.group_id = gu.group_id AND cg.permission_type IN (:group_permissions)
+        )
+        DELETE FROM user_chat_channel_memberships
+        WHERE (user_id, chat_channel_id) IN (
+          SELECT uccm.user_id, uccm.chat_channel_id
+          FROM user_chat_channel_memberships uccm
+          JOIN users u ON u.id = uccm.user_id AND u.id > 0 AND u.moderator = FALSE AND u.admin = FALSE #{user_sql}
```

Comment:
If a chat channel is only meant for admins, I think we need to auto remove moderators too? I'm not entirely sure because I don't have full context but it does seem odd if we don't automatically remove moderators.

Reply:
So I didn't change that behavior, just ported it over. It's not a _huge_ deal since the moderators won't have access to the chat channel because we check permissions in the controller. Though it's cleaner to also have the membership records removed. Will do that in another PR since this one is big enough as it is.